### PR TITLE
Support multitenancy in the URL Link sent for Devise Related operation

### PR DESCRIPTION
**Pull request by bivanalhar**
### Background

In our app, when a user:
- Sign Up
- Reset Password
- Request to Resend Confirmation Email

an automatic email with instructions is sent. These instructions typically contain a link for the user to click on to complete the next steps.

### Issue
The problem arises when user tried to do those operations from the tenant website (for example, `tenant1.coursemology.org`). 
1. The link provided on the email will direct user to the main website (`coursemology.org`) instead of the respective tenant website (`tenant1.coursemology.org`)
2. This behaviour confuses user due to the mismatch between the sites

### Root Cause

1. The process in sending the email for doing such operations was handled completely by the Devise gem
2. `default_url_options` are not being set, causing the gem to assume it as the base URL host

### Resolving the Issue

1. We add `parent_mailer` for Devise
2. Then, we setup the host for `default_url_options` towards the tenant address (in this case, it should be `tenant1.coursemology.org`)

### Additional Notes
For development and test environment in Rails, it's known that the site address is different from the production (by `2024/01/11`, the address used is `lvh.me` and the port was 8080 for development and 3200 for test). Therefore, the `default_url_options` will have its host value a bit different
- In case of development env, the link shall be `tenant1.lvh.me:8080`
- In case of test env, the link shall be `tenant1.lvh.me:3200`

### CODE REVIEW

File: app/mailers/devise_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+class DeviseMailer < ActionMailer::Base
+  before_action :set_mailers_url_host
+
+  DEFAULT_APP_HOST = 'lvh.me'
```
Comment (cysjonathan) [inline app/mailers/devise_mailer.rb:5]: Why are we hard-coding this instead of using `Application::Application.config.x.default_app_host`?

Comment (bivanalhar) [reply app/mailers/devise_mailer.rb]: refactored. noted on next time to be aware of this functionality

File: config/environments/test.rb [hunk 2]
```diff
@@ -57,8 +54,15 @@
...
   config.x.client_port = 3200
   config.x.server_port = 7979
+  config.x.default_app_host = "#{config.x.default_host}:3200"
   config.x.default_user_password = 'lolololol'
 
+  # the value should be 'lvh.me:3200'
+  config.x.initial_host = config.x.default_app_host
+
+  # We will assume that we are running on xxx.lvh.me:3200, where xxx is the tenant
+  config.action_mailer.default_url_options = { host: config.x.default_app_host }
```
Comment (cysjonathan) [inline config/environments/test.rb:64]: Why is your `default_url_options` in test.rb referencing `default_app_host` but in development.rb it is
`config.action_mailer.default_url_options = { host: config.x.default_host }`?

File: config/environments/test.rb [hunk 3]
```diff
@@ -57,8 +54,15 @@
   config.x.default_host = 'lvh.me'
   config.x.client_port = 3200
   config.x.server_port = 7979
+  config.x.default_app_host = "#{config.x.default_host}:3200"
```
Comment (cysjonathan) [inline config/environments/test.rb:57]: `config.x.default_app_host = "#{config.x.default_host}:#{config.x.client_port}"`
minimize use of magic numbers.
Your current code means that if you change your client port from 3200 to something else, you need to change two places, which may lead to oversight and bugs later on.
https://stackoverflow.com/questions/47882/what-are-magic-numbers-and-why-do-some-consider-them-bad

File: config/environments/test.rb (hunk 2 above)
Comment (bivanalhar) [reply config/environments/test.rb:64]: technically, the value of the host inside `default_url_option` should always be pointing towards `lvh.me:{port}` to make it consistent with the `development` environment.

However, the setup inside `development` and `test` for `default_app_host` and `default_host` are switched (inside `development`, it's stored inside `default_host` while inside `test`, it's stored inside `default_app_host`).

I think to make it consistent naming-wise, I'll refer to the `config.x.initial_host` for the `host` value inside `default_url_options`.

File: config/environments/test.rb (hunk 3 above)
Comment (bivanalhar) [reply config/environments/test.rb:57]: noted on this one..
also worth noting that previously in the `development` environment, the `client_port` over there was set to `5000`, which was the previous value used for this one. I will refactor this for more clarity

**Comment by ekowidianto** [comment]
Closing as this will be addressed when auth module is separated

**Pull request closed without merging**